File: GWT/src/cindyjs/quickhull3d/VertexList.java
package quickhull3d;

/**
 * Maintains a double-linked list of vertices for use by QuickHull3D
 */
class VertexList
{
	private Vertex head;
	private Vertex tail;

	/**
	 * Clears this list.
	 */
	public void clear()
	 {
	   head = tail = null; 
	 }

	/**
	 * Adds a vertex to the end of this list.
	 */
	public void add (Vertex vtx)
	 { 
	   if (head == null)
	    { head = vtx;
	    }
	   else
	    { tail.next = vtx; 
	    }
	   vtx.prev = tail;
	   vtx.next = null;
	   tail = vtx;
	 }

	/**
	 * Adds a chain of vertices to the end of this list.
	 */
	public void addAll (Vertex vtx)
	 { 
	   if (head == null)
	    { head = vtx;
	    }
	   else
	    { tail.next = vtx; 
	    }
	   vtx.prev = tail;
	   while (vtx.next != null)
	    { vtx = vtx.next;
	    }
	   tail = vtx;
	 }

	/**
	 * Deletes a vertex from this list.
	 */
	public void delete (Vertex vtx)
	 {
	   if (vtx.prev == null)
	    { head = vtx.next;
	    }
	   else
	    { vtx.prev.next = vtx.next; 
	    }
	   if (vtx.next == null)
	    { tail = vtx.prev; 
	    }
	   else
	    { vtx.next.prev = vtx.prev; 
	    }
	 }

	/**
	 * Deletes a chain of vertices from this list.
	 */
	public void delete (Vertex vtx1, Vertex vtx2)
	 {
	   if (vtx1.prev == null)
	    { head = vtx2.next;
	    }
	   else
	    { vtx1.prev.next = vtx2.next; 
	    }
	   if (vtx2.next == null)
	    { tail = vtx1.prev; 
	    }
	   else
	    { vtx2.next.prev = vtx1.prev; 
	    }
	 }

	/**
	 * Inserts a vertex into this list before another
	 * specificed vertex.
	 */
	public void insertBefore (Vertex vtx, Vertex next)
	 {
	   vtx.prev = next.prev;
	   if (next.prev == null)
	    { head = vtx;
	    }
	   else
	    { next.prev.next = vtx; 
	    }
	   vtx.next = next;
	   next.prev = vtx;
	 }

	/**
	 * Returns the first element in this list.
	 */
	public Vertex first()
	 {
	   return head;
	 }

	/**
	 * Returns true if this list is empty.
	 */
	public boolean isEmpty()
	 {
	   return head == null;
	 }
}

File: plugins/QuickHull3D/src/js/VertexList.js
var VertexList = function() {
    this.head = this.tail = null;
};

VertexList.prototype.clear = function() {
    this.head = this.tail = null;
};

VertexList.prototype.add = function(vertex) {
    if (this.head === null) {
        this.head = vertex;
    } else {
        this.tail.next = vertex;
    }

    vertex.previous = this.tail;
    vertex.next = null;
    this.tail = vertex;
};

VertexList.prototype.addAll = function(vertex) {
    if (this.head === null) {
        this.head = vertex;
    } else {
        this.tail.next = vertex;
    }

    vertex.previous = this.tail;

    while (vertex.next !== null) {
        vertex = vertex.next;
    }

    this.tail = vertex;
};

VertexList.prototype.delete = function(vertex1, vertex2) {
    if (vertex2 === void 0) {
        vertex2 = vertex1;
    }

    if (vertex1.previous === null) {
        this.head = vertex2.next;
    } else {
        vertex1.previous.next = vertex2.next;
    }

    if (vertex2.next === null) {
        this.tail = vertex1.previous;
    } else {
        vertex2.next.previous = vertex1.previous;
    }
};

VertexList.prototype.insertBefore = function(vertex, target) {
    vertex.previous = target.previous;

    if (target.previous === null) {
        this.head = vertex;
    } else {
        target.previous.next = vertex;
    }

    vertex.next = target;
    target.previous = vertex;
};

VertexList.prototype.isEmpty = function() {
    return this.head === null;
};

VertexList.prototype.length = function() {
    var result = 0;

    for (var vertex = this.head; vertex !== null; vertex = vertex.next) {
        result++;
    }

    return result;
};
